File: transform/Go/transformclient/transformclient.go
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// JSON, CSV, and XML transform helpers for Core Auto step scripts.

package transformclient

import (
	"bytes"
	"encoding/csv"
	"encoding/json"
	"encoding/xml"
	"fmt"
	"strings"

	"github.com/core-df/coreauto-scripts-pub/transform/Go/internal/result"
)

// JsonParse parses a JSON string into a Go value.
func JsonParse(text string) result.Result {
	var data any
	if err := json.Unmarshal([]byte(text), &data); err != nil {
		return result.Result{StatusCode: 400, Error: err.Error()}
	}
	return result.Result{StatusCode: 200, Data: data}
}

// JsonStringify serializes data to a JSON string.
func JsonStringify(data any, indent *int) result.Result {
	var out []byte
	var err error
	if indent != nil {
		out, err = json.MarshalIndent(data, "", strings.Repeat(" ", *indent))
	} else {
		out, err = json.Marshal(data)
	}
	if err != nil {
		return result.Result{StatusCode: 400, Error: err.Error()}
	}
	return result.Result{StatusCode: 200, Text: string(out)}
}

// CsvToRows parses CSV text into a slice of row maps.
func CsvToRows(text string, delimiter string) result.Result {
	if delimiter == "" {
		delimiter = ","
	}
	r := csv.NewReader(strings.NewReader(text))
	r.Comma = rune(delimiter[0])
	records, err := r.ReadAll()
	if err != nil {
		return result.Result{StatusCode: 400, Error: err.Error()}
	}
	if len(records) == 0 {
		return result.Result{StatusCode: 200, Rows: []map[string]string{}}
	}
	header := records[0]
	rows := make([]map[string]string, 0, len(records)-1)
	for _, rec := range records[1:] {
		row := make(map[string]string, len(header))
		for i, col := range header {
			if i < len(rec) {
				row[col] = rec[i]
			} else {
				row[col] = ""
			}
		}
		rows = append(rows, row)
	}
	return result.Result{StatusCode: 200, Rows: rows}
}

// RowsToCsv serializes row maps to CSV text.
func RowsToCsv(rows []map[string]string, delimiter string) result.Result {
	if len(rows) == 0 {
		return result.Result{StatusCode: 400, Error: "rows must not be empty"}
	}
	if delimiter == "" {
		delimiter = ","
	}
	var buf bytes.Buffer
	w := csv.NewWriter(&buf)
	w.Comma = rune(delimiter[0])

	fieldnames := make([]string, 0, len(rows[0]))
	for k := range rows[0] {
		fieldnames = append(fieldnames, k)
	}
	if err := w.Write(fieldnames); err != nil {
		return result.Result{StatusCode: 400, Error: err.Error()}
	}
	for _, row := range rows {
		record := make([]string, len(fieldnames))
		for i, col := range fieldnames {
			record[i] = row[col]
		}
		if err := w.Write(record); err != nil {
			return result.Result{StatusCode: 400, Error: err.Error()}
		}
	}
	w.Flush()
	if err := w.Error(); err != nil {
		return result.Result{StatusCode: 400, Error: err.Error()}
	}
	return result.Result{StatusCode: 200, Text: buf.String()}
}

type xmlNode struct {
	XMLName  xml.Name
	Content  string     `xml:",chardata"`
	Children []xmlNode  `xml:",any"`
	Attrs    []xml.Attr `xml:",any,attr"`
}

// XmlToDict parses XML text into a nested map keyed by root tag.
func XmlToDict(text string) result.Result {
	dec := xml.NewDecoder(strings.NewReader(text))
	var root xmlNode
	if err := dec.Decode(&root); err != nil {
		return result.Result{StatusCode: 400, Error: err.Error()}
	}
	data := map[string]any{root.XMLName.Local: elemToValue(root)}
	return result.Result{StatusCode: 200, Data: data}
}

func elemToValue(node xmlNode) any {
	if len(node.Children) == 0 {
		return strings.TrimSpace(node.Content)
	}
	out := make(map[string]any)
	for _, child := range node.Children {
		val := elemToValue(child)
		tag := child.XMLName.Local
		if existing, ok := out[tag]; ok {
			switch v := existing.(type) {
			case []any:
				out[tag] = append(v, val)
			default:
				out[tag] = []any{v, val}
			}
		} else {
			out[tag] = val
		}
	}
	return out
}

// DictToXml builds XML text from a map with the given root tag.
func DictToXml(data map[string]any, rootTag string) result.Result {
	if rootTag == "" {
		rootTag = "root"
	}
	root := &xmlElement{Name: rootTag}
	for k, v := range data {
		buildElement(root, v, k)
	}
	var buf bytes.Buffer
	buf.WriteString(xml.Header)
	enc := xml.NewEncoder(&buf)
	if err := enc.Encode(root); err != nil {
		return result.Result{StatusCode: 400, Error: err.Error()}
	}
	text := strings.TrimSpace(buf.String())
	if strings.HasPrefix(text, "<?xml") {
		if idx := strings.Index(text, "?>"); idx >= 0 {
			text = strings.TrimSpace(text[idx+2:])
		}
	}
	return result.Result{StatusCode: 200, Text: text}
}

type xmlElement struct {
	XMLName  xml.Name
	Name     string       `xml:"-"`
	Content  string       `xml:",chardata"`
	Children []*xmlElement `xml:",any"`
}

func (e *xmlElement) MarshalXML(enc *xml.Encoder, start xml.StartElement) error {
	tag := e.Name
	if tag == "" {
		tag = e.XMLName.Local
	}
	start.Name = xml.Name{Local: tag}
	if err := enc.EncodeToken(start); err != nil {
		return err
	}
	if len(e.Children) == 0 {
		if err := enc.EncodeToken(xml.CharData(e.Content)); err != nil {
			return err
		}
	} else {
		for _, child := range e.Children {
			if err := enc.Encode(child); err != nil {
				return err
			}
		}
	}
	return enc.EncodeToken(start.End())
}

func buildElement(parent *xmlElement, obj any, tag string) {
	switch v := obj.(type) {
	case map[string]any:
		node := &xmlElement{Name: tag}
		parent.Children = append(parent.Children, node)
		for k, val := range v {
			buildElement(node, val, k)
		}
	case []any:
		for _, item := range v {
			buildElement(parent, item, tag)
		}
	default:
		text := ""
		if obj != nil {
			text = fmt.Sprint(obj)
		}
		parent.Children = append(parent.Children, &xmlElement{Name: tag, Content: text})
	}
}

File: transform/Rust/src/transformclient.rs
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use quick_xml::events::Event;
use quick_xml::reader::Reader;
use quick_xml::Writer;
use serde_json::{json, Map, Value};
use std::io::Cursor;

pub fn json_parse(text: &str) -> Value {
    match serde_json::from_str::<Value>(text) {
        Ok(data) => json!({ "status_code": 200, "data": data }),
        Err(e) => json!({ "status_code": 400, "error": e.to_string() }),
    }
}

pub fn json_stringify(data: &Value, indent: Option<usize>) -> Value {
    let res = if let Some(n) = indent {
        serde_json::to_string_pretty(data)
    } else {
        serde_json::to_string(data)
    };
    match res {
        Ok(text) => json!({ "status_code": 200, "text": text }),
        Err(e) => json!({ "status_code": 400, "error": e.to_string() }),
    }
}

fn split_csv_line(line: &str, delimiter: char) -> Vec<String> {
    let mut fields = Vec::new();
    let mut cur = String::new();
    let mut in_quotes = false;
    for ch in line.chars() {
        match ch {
            '"' if !in_quotes => in_quotes = true,
            '"' if in_quotes => in_quotes = false,
            c if c == delimiter && !in_quotes => {
                fields.push(cur.clone());
                cur.clear();
            }
            c => cur.push(c),
        }
    }
    fields.push(cur);
    fields
}

pub fn csv_to_rows(text: &str, delimiter: &str) -> Value {
    let delim = delimiter.chars().next().unwrap_or(',');
    let mut lines = text.lines();
    let header = match lines.next() {
        Some(h) => h,
        None => return json!({ "status_code": 400, "error": "empty csv" }),
    };
    let cols: Vec<String> = split_csv_line(header, delim);
    let mut rows = Vec::new();
    for line in lines {
        if line.trim().is_empty() {
            continue;
        }
        let vals = split_csv_line(line, delim);
        let mut row = Map::new();
        for (i, col) in cols.iter().enumerate() {
            let v = vals.get(i).cloned().unwrap_or_default();
            row.insert(col.clone(), Value::String(v));
        }
        rows.push(Value::Object(row));
    }
    json!({ "status_code": 200, "rows": rows })
}

pub fn rows_to_csv(rows: &[Map<String, Value>], delimiter: &str) -> Value {
    if rows.is_empty() {
        return json!({ "status_code": 400, "error": "rows must not be empty" });
    }
    let delim = delimiter.chars().next().unwrap_or(',');
    let cols: Vec<String> = rows[0].keys().cloned().collect();
    let mut out = cols.join(&delim.to_string());
    out.push('\n');
    for row in rows {
        let parts: Vec<String> = cols
            .iter()
            .map(|c| row.get(c).map(value_to_string).unwrap_or_default())
            .collect();
        out.push_str(&parts.join(&delim.to_string()));
        out.push('\n');
    }
    json!({ "status_code": 200, "text": out })
}

fn value_to_string(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        Value::Null => String::new(),
        other => other.to_string(),
    }
}

fn xml_elem_to_value(reader: &mut Reader<Cursor<&[u8]>>, name: &[u8]) -> Result<Value, String> {
    let mut children: Map<String, Value> = Map::new();
    let mut text = String::new();
    loop {
        match reader.read_event().map_err(|e| e.to_string())? {
            Event::Start(e) => {
                let tag = String::from_utf8_lossy(e.name().as_ref()).to_string();
                let val = xml_elem_to_value(reader, e.name().as_ref())?;
                merge_child(&mut children, tag, val);
            }
            Event::Text(t) => {
                text.push_str(&String::from_utf8_lossy(&t.into_inner()));
            }
            Event::End(e) if e.name().as_ref() == name => {
                if children.is_empty() {
                    return Ok(Value::String(text.trim().to_string()));
                }
                return Ok(Value::Object(children));
            }
            Event::Eof => return Err("unexpected eof".into()),
            _ => {}
        }
    }
}

fn merge_child(map: &mut Map<String, Value>, tag: String, val: Value) {
    if let Some(existing) = map.remove(&tag) {
        let arr = match existing {
            Value::Array(mut a) => {
                a.push(val);
                a
            }
            other => vec![other, val],
        };
        map.insert(tag, Value::Array(arr));
    } else {
        map.insert(tag, val);
    }
}

pub fn xml_to_dict(text: &str) -> Value {
    let mut reader = Reader::from_str(text);
    reader.config_mut().trim_text(true);
    loop {
        match reader.read_event() {
            Ok(Event::Start(e)) => {
                let tag = String::from_utf8_lossy(e.name().as_ref()).to_string();
                match xml_elem_to_value(&mut reader, e.name().as_ref()) {
                    Ok(inner) => {
                        let mut data = Map::new();
                        data.insert(tag, inner);
                        return json!({ "status_code": 200, "data": Value::Object(data) });
                    }
                    Err(e) => return json!({ "status_code": 400, "error": e }),
                }
            }
            Ok(Event::Eof) => break,
            Err(e) => return json!({ "status_code": 400, "error": e.to_string() }),
            _ => {}
        }
    }
    json!({ "status_code": 400, "error": "empty xml" })
}

fn write_xml_value(writer: &mut Writer<Cursor<Vec<u8>>>, tag: &str, val: &Value) -> Result<(), String> {
    match val {
        Value::Object(map) => {
            writer
                .create_element(tag)
                .with_attributes([])
                .write_inner_content(|w| {
                    for (k, v) in map {
                        write_xml_value(w, k, v)?;
                    }
                    Ok(())
                })
                .map_err(|e| e.to_string())?;
        }
        Value::Array(arr) => {
            for item in arr {
                write_xml_value(writer, tag, item)?;
            }
        }
        other => {
            let text = value_to_string(other);
            writer
                .create_element(tag)
                .with_text(text.as_bytes())
                .map_err(|e| e.to_string())?;
        }
    }
    Ok(())
}

pub fn dict_to_xml(data: &Map<String, Value>, root_tag: &str) -> Value {
    let mut writer = Writer::new(Cursor::new(Vec::new()));
    if let Err(e) = writer
        .create_element(root_tag)
        .write_inner_content(|w| {
            for (k, v) in data {
                write_xml_value(w, k, v)?;
            }
            Ok(())
        })
    {
        return json!({ "status_code": 400, "error": e.to_string() });
    }
    let buf = writer.into_inner().into_inner();
    match String::from_utf8(buf) {
        Ok(text) => json!({ "status_code": 200, "text": text }),
        Err(e) => json!({ "status_code": 400, "error": e.to_string() }),
    }
}
